### PR TITLE
Linux Support (again)

### PR DESCRIPTION
**Pull request by fearphage**
This is the fixes (and some) of the PR #7. I couldn't figure out how to continue progress on that branch/PR.

Let me know if everything still works.

Question: Why don't we/you use `ffmpeg` on all platforms? It seems like it would greatly simplify the code and it would add support for MacOS older than 10.10.

### CODE REVIEW

**Comment by sindresorhus** [comment]
> Why don't we/you use ffmpeg on all platforms?

https://github.com/wulkano/aperture#why

File: index.js [hunk 1]
```diff
@@ -9,11 +9,17 @@ const debuglog = util.debuglog('aperture');
...
+      macosVersion.assertGreaterThanOrEqualTo('10.10');
+    }
   }
 
   getAudioSources() {
-    return execa.stdout(path.join(__dirname, 'swift/main'), ['list-audio-devices']).then(JSON.parse);
+    if (process.platform === 'darwin') {
+      return execa.stdout(path.join(__dirname, 'swift/main'), ['list-audio-devices']).then(JSON.parse);
+    } else if (process.platform === 'linux') {
+      return execa.stdout('sh', ['-c', "arecord -l | awk 'match(\$0, /card ([0-9]): ([^,]+),/, result) { print result[1] \":\" result[2] }'"]);
```
Comment (sindresorhus) [inline index.js]: We should use JS for as much as possible. No awk.

File: package.json [hunk 2]
```diff
@@ -17,8 +17,9 @@
   },
   "scripts": {
     "test": "xo",
-    "build": "cd swift && xcodebuild -derivedDataPath $(mktemp -d) -scheme aperture",
-    "prepublish": "npm run build"
+    "build-macos": "cd swift && xcodebuild -derivedDataPath $(mktemp -d) -scheme aperture",
+    "postinstall": "[ \"$(uname)\" = \"Darwin\" ] && npm run build-macos",
```
Comment (sindresorhus) [inline package.json]: This one should be removed.

**Comment by sindresorhus** [comment]
Also note my comment in #7: https://github.com/wulkano/aperture/pull/7#issuecomment-265879006

File: index.js [hunk 3]
```diff
@@ -9,11 +9,17 @@ const debuglog = util.debuglog('aperture');
 
 class Aperture {
   constructor() {
-    macosVersion.assertGreaterThanOrEqualTo('10.10');
+    if (process.platform === 'darwin') {
```
Comment (sindresorhus) [inline index.js]: Create top level constants IS_MACOS and IS_LINUX to simplify all these if else statements.

**Comment by sindresorhus** [comment]
> Let me know if everything still works.

You can do that by putting the following in `index.js`:

```js
module.exports().startRecording().then(console.log).catch(console.error);
```

**Comment by fearphage** [comment]
> You can do that by putting the following

I'm sure it works for me. I just have zero macs on which to confirm the functionality.

**Comment by fearphage** [comment]
Not sure what changed between last night and today, but I was getting `string`s back from the `stderr` stream and now I'm getting `buffer`s. Weird.

**Comment by fearphage** [comment]
@sindresorhus Anything else?

File: index.js [hunk 4]
```diff
@@ -45,20 +66,37 @@ class Aperture {
...
+        if (typeof cropArea === 'object') {
+          args.push(
+            '-video_size', `${cropArea.width}x${cropArea.height}`,
+            '-i', `:0+${cropArea.x},${cropArea.y}`
+          );
+        } else {
+          args.push('-i', ':0');
+        }
+
+        args.push('-framerate', fps, '-draw_mouse', +(showCursor === true), this.tmpPath);
```
Comment (sindresorhus) [inline index.js]: `String(showCursor === true)`

**Comment by sindresorhus** [comment]
Setting `showCursor: false` doesn't seem to work. I still get the cursor in the recording.

By default it captures only part of the screen, while it should capture the whole screen. Setting the `cropArea` option does nothing.

`.getAudioSources()` return duplicate audio sources for me:

```
[ '0:I82801BAICH2 [Intel 82801BA-ICH2]',
  '0:I82801BAICH2 [Intel 82801BA-ICH2]' ]
```

---

ffmpeg version 3.0.7-0ubuntu0.16.10.1

File: index.js (hunk 4 above)
Comment (fearphage) [inline index.js]: Did you verify this works? [Based on the documentation](https://ffmpeg.org/ffmpeg-devices.html#Options-19), `draw_mouse` expects a binary value. That's why I'm coercing the boolean to an int.

**Comment by fearphage** [comment]
> Setting showCursor: false doesn't seem to work

Confirmed.


> By default it captures only part of the screen, while it should capture the whole screen

I'll look into it.


> .getAudioSources() return duplicate audio sources for me

Cannot reproduce locally.


> It also needs to be in the format

From your sample output, is that `id [name]`?


> ffmpeg version 3.0.7-0ubuntu0.16.10.1

Same here.

File: index.js [hunk 5]
```diff
@@ -7,13 +7,32 @@ const macosVersion = require('macos-version');
...
+      macosVersion.assertGreaterThanOrEqualTo('10.10');
+    }
   }
 
   getAudioSources() {
-    return execa.stdout(path.join(__dirname, 'swift/main'), ['list-audio-devices']).then(JSON.parse);
+    if (IS_MACOS) {
+      return execa.stdout(path.join(__dirname, 'swift/main'), ['list-audio-devices']).then(JSON.parse);
+    } else if (IS_LINUX) {
+      return execa.stdout('arecord', ['-l']).then(
```
Comment (sindresorhus) [inline index.js]: `-l` => `--list-devices`

**Comment by sindresorhus** [comment]
> From your sample output, is that id [name]?

```
$ arecord --list-devices
**** List of CAPTURE Hardware Devices ****
card 0: I82801BAICH2 [Intel 82801BA-ICH2], device 0: Intel ICH [Intel 82801BA-ICH2]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: I82801BAICH2 [Intel 82801BA-ICH2], device 1: Intel ICH - MIC ADC [Intel 82801BA-ICH2 - MIC ADC]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```

**Comment by sindresorhus** [comment]
It seems if you specify `default` to ffmpeg, it will use the default device. According to the web it's sometimes hard to determine this yourself, so maybe we should add in a `Default Microphone` entry, that just supplies `-i default` to ffmpeg?

**Pull request closed without merging**